### PR TITLE
Use PlugAttack

### PR DESCRIPTION
**Pull request by michalmuskala**
This is a PoC for replacing custom components for rate limiting and ip blocking with `plug_attack` package, I recently released.

It has couple advantages:
* increasing rate limit for authenticated users is trivial - just change the `limit` option based on some connection params,
* skipping rate limiting of conditional requests should be easier to do with a plug-like interface (although we know about that only after the controller processes the request, so it might require some additional work)
* there is a clean storage interface, so implementing something with redis should be easier
* I'm considering adding a loadable blocklist - this should simplify the `BlockAddress` thing even more.

### CODE REVIEW

File: web/plugs/plug_attack.ex [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    |> put_resp_header("x-ratelimit-remaining", Integer.to_string(data[:remaining]))
+    |> put_resp_header("x-ratelimit-reset", Integer.to_string(reset))
+  end
+
+  defp ip_str({a, b, c, d}) do
+    "#{a}.#{b}.#{c}.#{d}"
+  end
+
+  defp check_blocked(ip) do
+    case :ets.lookup(@addresses_ets, ip) do
```
Comment (ericmj) [inline web/plugs/plug_attack.ex]: This lookup and the one below should be functions on BlockAddress instead. We shouldn't use the ets table outside of that module.

**Comment by ericmj** [comment]
What would look like if we used functions instead of using the DSL?

File: web/plugs/plug_attack.ex (hunk 1 above)
Comment (michalmuskala) [inline web/plugs/plug_attack.ex]: That is 100% right.

**Comment by michalmuskala** [comment]
The `rule` macro defines a function with the same name and registers the function to be called in `call` callback of the plug.

Something like:
```elixir
rule "foo", conn, do: ...
rule "bar", conn, do: ...
```
Translates to:
```elixir
def foo(conn), do: ...
def bar(conn), do: ...

def call(conn, opts) do
  case foo(conn) do
    {:allow, data} -> allow_action(conn, data, opts)
    {:block, data} -> block_action(conn, data, opts)
    nil ->
      case bar(conn) do
        {:allow, data} -> allow_action(conn, data, opts)
        {:block, data} -> block_action(conn, data, opts)
        nil -> conn
  end
end
```

**Comment by ericmj** [comment]
Okay, then I think it makes sense to use the DSL, but it's worth investigating if there is a way to make it work nicely with just functions.

BTW, is the rule name used for the function name only or is it used for something else also?

**Comment by ericmj** [comment]
Can we use plug_attack from hex and resolve the conflict in mix.lock? Should be good to merge after that.

**Comment by ericmj** [comment]
Ping? :)

**Comment by michalmuskala** [comment]
Sorry, I will look into this today.

**Comment by wojtekmach** [comment]
@michalmuskala ugh, sorry I've updated mix.{exs,lock} in the meantime, do you mind rebasing and using a released version on hex? Looks great to me, btw!

**Comment by ericmj** [comment]
Ping? :)

**Comment by ericmj** [comment]
@michalmuskala Is it okay if I finish this if you don't have time?

**Comment by michalmuskala** [comment]
@ericmj Updated.

I'm really sorry it too so embarrassingly long.

**Comment by ericmj** [comment]
No worries. Thank you! :heart:

**Pull request merged**